### ISSUE TITLE
Incompatibility with composer ~1.3.0 and setup:di:compile

### ISSUE THREAD

**Issue by AntonEvers** [magento/magento2#8635]
### Preconditions

1. Install composer 1.3.0 or higher
2. Check out Magento 2.0.10

### Steps to reproduce

1. run composer update
2. run bin/magento setup:di:compile-multi-tenant -vvv

### Expected result

1. compilation runs fine

### Actual result

  [Magento\Framework\Exception\LocalizedException]
  Source class "\Magento\Catalog\Model\Product\OptionInterface" for "\Magento\Catalog\Model\Product\OptionInterfacePersistor" genera
  tion does not exist.

I've added the solution to this in https://magento.stackexchange.com/questions/147219/optioninterfacepersistor-error-when-compiling-magento-in-production-mode/161010#161010

I will create a pull request for you to fix this.

**Comment by orlangur** [comment]
> run composer update

Does not seem to be a totally safe command to me. Will `./vendor/bin/composer update` do the trick?

As to me, it's better to have Composer 1.3.x completely supported (see https://github.com/magento/magento2/issues/4359) than do some magic like this one.

**Comment by AntonEvers** [comment]
Hello @orlangur,

I wasn't aware of ticket https://github.com/magento/magento2/issues/4359.
Let's keep this ticket pending until #4359 is fixed indeed.

As to the recommended composer executable, http://devdocs.magento.com/guides/v2.1/install-gde/prereq/integrator_install_composer.html will lead more developers to using composer 1.3.x. 

In my case `./vendor/composer/composer/bin/composer update` certainly does the trick. `./vendor/bin/composer update` gives me errors.

```
PHP Warning:  require(/var/www/vhosts/magento2/httpdocs/vendor/bin/../src/bootstrap.php): failed to open stream: No such file or directory in /var/www/vhosts/magento2/httpdocs/vendor/bin/composer on line 8
```

Using `./vendor/composer/composer/bin/composer` looks like a promising workaround for now.

**Comment by orlangur** [comment]
@ajpevers, yeah, I checked existing documentation before writing comment and there is no suggestion to use `composer` from `vendor`.

Thanks, good to know that `./vendor/bin/composer` is not working.

The bad thing with relying on unstable version is that in some cases both of them could not work. I faced with such issues when adding new requirements - in some Composer versions there were bugs with version contraint resolution.
